app/position/providers/androidpositionprovider.cpp

```diff
@@ -0,0 +1,277 @@
+  {
+    const jfloat bearing = location.callMethod<jfloat>( "getBearing" );
+    if ( !qFuzzyIsNull( bearing ) )
+      pos.direction = bearing;
+
+    // could also use getBearingAccuracyDegrees() since API level 26 (Android 8.0)
+  }
+
+  // could also use isMock() to detect if location is mocked
+  // (may useful to check if 3rd party app is setting it for external GNSS receiver)
+
+  // could also use getExtras() to get further details from mocked location
+  // (the key/value pairs are vendor-specific, and could include things like DOP,
+  // info about corrections, geoid undulation, receiver model)
+
+  __android_log_print( ANDROID_LOG_INFO, "CPP", "[c++] pos %f %f", latitude, longitude );
+
+  QJniObject gnssStatus( gnssStatusObj );
+  if ( gnssStatus.isValid() )
+  {
+    int satellitesUsed = 0;
+    const int satellitesCount = gnssStatus.callMethod<jint>( "getSatelliteCount" );
+    for ( int i = 0; i < satellitesCount; ++i )
+    {
+      if ( gnssStatus.callMethod<jboolean>( "usedInFix", i ) )
+        ++satellitesUsed;
+
+      // we could get more info here (ID, azimuth, elevation, signal strength, ...)
+      // but we are not using that anywhere
+    }
+
+    pos.satellitesVisible = satellitesCount;
+    pos.satellitesUsed = satellitesUsed;
+  }
+
+  QMetaObject::invokeMethod( inst, "positionChanged",
+                             Qt::AutoConnection, Q_ARG( GeoPosition, pos ) );
+
+}
+
+
+AndroidPositionProvider::AndroidPositionProvider( bool fused, QObject *parent )
+  : AbstractPositionProvider( fused ? QStringLiteral( "android_fused" ) : QStringLiteral( "android_gps" ),
+                              QStringLiteral( "internal" ),
+                              fused ? tr( "Internal (fused)" ) : tr( "Internal (gps)" ), parent )
+  , mFused( fused )
+  , mInstanceId( ++sLastInstanceId )
+{
+  __android_log_print( ANDROID_LOG_INFO, "CPP", "[c++] CONSTRUCT" );
+
+  Q_ASSERT( !sInstances.contains( mInstanceId ) );
+  sInstances[mInstanceId] = this;
+
+  // register the native methods
+
+  JNINativeMethod methods[]
+  {
+    {
+      "jniOnPositionUpdated",
+      "(ILandroid/location/Location;Landroid/location/GnssStatus;)V",
+      reinterpret_cast<void *>( jniOnPositionUpdated )
+    }
+  };
+
+  QJniEnvironment javaenv;
+
+  javaenv.registerNativeMethods( "uk/co/lutraconsulting/MMAndroidPosition", methods, 1 );
+
+  // create the Java object
+
+  __android_log_print( ANDROID_LOG_INFO, "CPP", "[c++] create Java object" );
+
+  jobject context = QNativeInterface::QAndroidApplication::context();
+
+  mAndroidPos = QJniObject::callStaticObjectMethod( "uk/co/lutraconsulting/MMAndroidPosition", "createWithJniCallback",
+                "(Landroid/content/Context;ZI)Luk/co/lutraconsulting/MMAndroidPosition;", context, mFused, mInstanceId );
+
+  // Request permissions if needed
+
+  QLocationPermission perm;
+  perm.setAccuracy( QLocationPermission::Precise );
+  if ( qApp->checkPermission( perm ) != Qt::PermissionStatus::Granted )
```

Comment:
Will this now prompt for permissions again at app startup? It currently prompts user when a project is opened.

Reply:
Fixed in adfa992 ... (it was also causing a crash when prompting permissions upon start)